library/std/src/io/buffered/bufreader.rs

```diff
@@ -94,6 +94,32 @@ impl<R: Read> BufReader<R> {
     pub fn with_capacity(capacity: usize, inner: R) -> BufReader<R> {
         BufReader { inner, buf: Buffer::with_capacity(capacity) }
     }
+
+    /// Attempt to look ahead `n` bytes.
+    ///
+    /// `n` must be less than `capacity`.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// #![feature(bufreader_peek)]
+    /// use std::io::{Read, BufReader};
+    ///
+    /// let mut bytes = &b"hello"[..];
+    /// let mut rdr = BufReader::with_capacity(6, &mut bytes);
+    /// assert_eq!(rdr.peek(2).unwrap(), b"he");
+    /// let mut s = String::new();
+    /// rdr.read_to_string(&mut s).unwrap();
+    /// assert_eq!(&s, "hello");
+    /// ```
+    #[unstable(feature = "bufreader_peek", issue = "128405")]
+    pub fn peek(&mut self, n: usize) -> io::Result<&[u8]> {
+        assert!(n < self.capacity());
+        while n > self.buf.buffer().len() {
+            self.buf.read_more(&mut self.inner)?;
+        }
```

Comment:
`buf` can have the current data sitting near the end, so the overall capacity of the buffer minus the length is not the remaining tail capacity.

So even if `n < capacity` it doesn't mean that amount can be read. So a backshift can be necessary.

You should also add internal tests covering that case.